### ISSUE TITLE
Investigate `datadog_logs -> datadog_logs` performance. 

### ISSUE THREAD

**Issue by blt** [timberio/vector#8263]
This issue is a record of and parent to the investigation of vector's performance when configured like so:

```toml
##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.dd_logs]
type = "datadog_logs" 
acknowledgements = false 
address = "0.0.0.0:8282" 

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["dd_logs"]
default_api_key = "${DATADOG_API_KEY_ENV_VAR}"
encoding.codec = "json"
request.concurrency = "adaptive"
batch.max_bytes = 5242880
request.rate_limit_num = 1000
```

Throughput for this setup is lower than desired, roughly 1.5Mb/s on the latest nightly with payloads into `dd_logs` ranging from 2Mb to 4Kb, as generated by [`http_gen`](https://github.com/blt/lading/tree/main/http_gen) with the following configuration:

```toml
worker_threads = 2
prometheus_addr = "0.0.0.0:9001"

[targets.simple]
target_uri = "http://localhost:8282/v1/input"
bytes_per_second = "100 Mb"
parallel_connections = 25
method.type = "Post"
method.variant = "DatadogLog"
method.maximum_prebuild_cache_size_bytes = "500 Mb"

[targets.simple.headers]
dd-api-key = "{{ datadog_api_key }}"
```

Captured pprof data -- see [6633e35e-b912-4173-94d1-a7b817e44dc8.pprof.zip](https://github.com/timberio/vector/files/6810717/6633e35e-b912-4173-94d1-a7b817e44dc8.pprof.zip) -- suggests that the bulk of CPU time is spend in compression related activities, as captured by our soak environment.

**Comment by blt** [comment]
I've started to run a simplified vector build:

```
> CARGO_PROFILE_RELEASE_DEBUG=2 cargo build --release --no-default-features --features "sources-datadog,sinks-blackhole,sources-internal_metrics,sinks-prometheus"
```
with this configuration:

```toml
##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs" # required
acknowledgements = false # optional, default
address = "0.0.0.0:8282" # required

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.blackhole]
type = "blackhole"
inputs = ["datadog_logs"]
```

Throughput is better, as expected, at just over 50Mb/s on my system, meaning that the sink is a penalty of 48Mb/s. Same `http_gen` configuration as the prior comment. Notable cost areas:

* The use of `serde_json::to_string` in the blackhole sink to get byte sizes 
* The use of `vector::sources::util::decode_body` to decode incoming payloads, secondarily. 
* The path iteration code shows up as well, as discussed more fully in #8215 

The blackhole is, of course, not part of the investigation chain but its bottleneck bottlenecks the upstream sender.

**Comment by blt** [comment]
Excising any of the json serialization in blackhole sink made throughput climb to 80Mb/s. 

```
diff --git a/src/sinks/blackhole.rs b/src/sinks/blackhole.rs
index 8668b6a4e..c9d2fcc81 100644
--- a/src/sinks/blackhole.rs
+++ b/src/sinks/blackhole.rs
@@ -89,24 +89,24 @@ impl StreamSink for BlackholeSink {
                 self.last = Some(until);
             }

-            let message_len = match event {
-                Event::Log(log) => serde_json::to_string(&log),
-                Event::Metric(metric) => serde_json::to_string(&metric),
-            }
-            .map(|v| v.len())
-            .unwrap_or(0);
+            // let message_len = match event {
+            //     Event::Log(log) => serde_json::to_string(&log),
+            //     Event::Metric(metric) => serde_json::to_string(&metric),
+            // }
+            // .map(|v| v.len())
+            // .unwrap_or(0);

             self.total_events += 1;
-            self.total_raw_bytes += message_len;
+            // self.total_raw_bytes += message_len;

-            emit!(BlackholeEventReceived {
-                byte_size: message_len
-            });
+            // emit!(BlackholeEventReceived {
+            //     byte_size: message_len
+            // });

             if self.total_events % self.config.print_amount == 0 {
                 info!({
                     events = self.total_events,
-                    raw_bytes_collected = self.total_raw_bytes
+                    // raw_bytes_collected = self.total_raw_bytes
                 }, "Total events collected");
             }
```

**Comment by blt** [comment]
With this previous change metrics/tracing costs come up in the world. Notably `t1ha` costs are now 6% of overall runtime, which is removed in upstream metrics. Vector is currently one point release behind.

**Comment by blt** [comment]
Upgrading metrics to the latest releases improved throughput to 120Mb/s. 

```
diff --git a/Cargo.toml b/Cargo.toml
index 4205b8d36..ea522f85a 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -127,9 +127,9 @@ opentelemetry = { version = "0.14.0", default-features = false, features = ["tra
 opentelemetry-datadog = { version = "0.2.0", default-features = false, features = ["reqwest-client"] }

 # Metrics
-metrics = { version = "0.16.0", default-features = false, features = ["std"] }
+metrics = { version = "0.17.0", default-features = false, features = ["std"] }
 metrics-tracing-context = { version = "0.6.0", default-features = false }
-metrics-util = { version = "0.9.1", default-features = false, features = ["std"] }
+metrics-util = { version = "0.10.0", default-features = false, features = ["std"] }
```

**Comment by blt** [comment]
I've gained another 20Mb/s with this diff, leading to a consistent 140Mb/s:

```
diff --git a/src/sources/datadog/logs.rs b/src/sources/datadog/logs.rs
index 28409ea1d..2d43f1b01 100644
--- a/src/sources/datadog/logs.rs
+++ b/src/sources/datadog/logs.rs
@@ -1,20 +1,21 @@
+use crate::sources::util::{ErrorMessage, HttpSource, HttpSourceAuthConfig};
 use crate::{
     config::{
         log_schema, DataType, GenerateConfig, Resource, SourceConfig, SourceContext,
         SourceDescription,
     },
     event::Event,
-    sources::{
-        self,
-        util::{decode_body, Encoding, ErrorMessage, HttpSource, HttpSourceAuthConfig},
-    },
+    sources,
     tls::TlsConfig,
 };
 use bytes::Bytes;
+use chrono::Utc;
 use lazy_static::lazy_static;
 use regex::Regex;
 use serde::{Deserialize, Serialize};
 use std::{collections::HashMap, net::SocketAddr, sync::Arc};
+use vector_core::event::LogEvent;
+use warp::http::StatusCode;

 use warp::http::HeaderMap;

@@ -77,6 +78,12 @@ struct DatadogLogsSource {
     store_api_key: bool,
 }

+#[derive(Deserialize)]
+struct LogMsg {
+    pub message: String,
+    pub timestamp: i64,
+}
+
 impl HttpSource for DatadogLogsSource {
     fn build_events(
         &self,
@@ -86,7 +93,7 @@ impl HttpSource for DatadogLogsSource {
         request_path: &str,
     ) -> Result<Vec<Event>, ErrorMessage> {
         if body.is_empty() {
-            // The datadog agent may sent empty payload as keep alive
+            // The datadog agent may send an empty payload as a keep alive
             debug!(
                 message = "Empty payload ignored.",
                 internal_log_rate_secs = 30
@@ -100,18 +107,28 @@ impl HttpSource for DatadogLogsSource {
         }
         .map(Arc::from);

-        decode_body(body, Encoding::Json).map(|mut events| {
-            // Datadog API key in metadata & source type field
-            let key = log_schema().source_type_key();
-            for event in &mut events {
-                let log = event.as_mut_log();
+        let messages: Vec<LogMsg> = serde_json::from_slice(&body).map_err(|error| {
+            ErrorMessage::new(
+                StatusCode::BAD_REQUEST,
+                format!("Error parsing JSON: {:?}", error),
+            )
+        })?;
+        messages
+            .into_iter()
+            .map(|msg| {
+                let mut log = LogEvent::default();
+                log.insert(log_schema().timestamp_key(), Utc::now()); // Add timestamp
+                log.insert_flat("message".to_string(), msg.message);
+                log.insert_flat("timestamp".to_string(), msg.timestamp);
+                let key = log_schema().source_type_key();
                 log.try_insert(key, Bytes::from("datadog_logs"));
                 if let Some(k) = &api_key {
                     log.metadata_mut().set_datadog_api_key(Some(Arc::clone(k)));
                 }
-            }
-            events
-        })
+                log
+            })
+            .map(|log| Ok(log.into()))
+            .collect()
     }
 }
```

Instead of routing through the generic `decode_body` we directly deserialize the logs message -- presumed -- and convert to a `Log` from there. Notably performance is much more consistent than in the previous commit. 

It's not known to me -- paging @prognant  -- if the `LogMsg` here accurately reflects the payload the datadog agent will send.

**Comment by blt** [comment]
Profiling suggested `LogEvent::contains` as a choke point, consider how often we insert in this workflow.  #8331 addresses this by avoiding path iteration if the key is present in the top-level of the fields tree, granting +2Mb/s. Performance is unchanged if the field tree is deep and the path is as well, though we might consider alternative representations to improve that. This will strongly depend on what the path iteration language is, which I'm unsure of.

**Comment by blt** [comment]
`atomic_load` is now 5.9% of total execution cycles, with all the changes presented above layered up. The primary caller we can do something about is https://github.com/timberio/vector/blob/90df0c6f6f92b429da8b3a46873c55ef5723acc7/lib/vector-core/src/config/log_schema.rs#L41-L43

`LOG_SCHEMA` there is a `lazy_static!` `sync::OnceCell` so that's where the atomics come in.

**Comment by blt** [comment]
> `atomic_load` is now 5.9% of total execution cycles, with all the changes presented above layered up. The primary caller we can do something about is
> 
> https://github.com/timberio/vector/blob/90df0c6f6f92b429da8b3a46873c55ef5723acc7/lib/vector-core/src/config/log_schema.rs#L41-L43
> 
> `LOG_SCHEMA` there is a `lazy_static!` `sync::OnceCell` so that's where the atomics come in.

This is adjusted in https://github.com/timberio/vector/pull/8297/commits/f1149cea183eccfd876fd0eb86fc4f8ce165749f, I have also removed the call to `try_insert` which removes the immediate need for https://github.com/timberio/vector/pull/8331. The change here is worth +100Mb/s, mostly by allowing more parallel execution. 

At this point my working branch is a swiss-cheese of rebasing. I will repeat my measurements once changes land upstream to get a clearer sense of things.

**Comment by blt** [comment]
Alright. I've now broken the major changed up into either branches or actual PRs. As a reminder the current vector configuration is:

```toml
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs" # required
acknowledgements = false # optional, default
address = "0.0.0.0:8282" # required

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.blackhole]
type = "blackhole"
inputs = ["datadog_logs"]
```

SHA 529dfbf422a3a9682e11991b34fdfde9a28e1066 is the baseline measure here, with vector limited to 8 CPUs on "AMD Ryzen Threadripper PRO 3955WX". Linux frequency set to 'schedutil' for all system cores. The new baseline is 165Mb/s. This is noticeably faster than previous measurements taken [here](https://github.com/timberio/vector/issues/8263#issuecomment-879405832) for which I have no sound hypothesis, as the SHA baseline was not recorded. That said, these measurements are _NOT_ taken while running under `perf`. 

| Summary | PR/commit | Change Mb/s over baseline |
| ------------- | --------------  | -----------------------   |
| blackhole with no bytes processed | https://github.com/timberio/vector/commit/f39e725e132dffe13aeb087af56c4da7a825c02d | +155 |
| faster DD agent parsing  | https://github.com/timberio/vector/pull/8297 | +12 |
| experimental metrics update | https://github.com/timberio/vector/commit/28d6a2e9020ab10834f602a6c6eb683ff734fda3 | +18 |
| faster `LogEvent::contains` | https://github.com/timberio/vector/pull/8331 | +3 |
| all of the above | https://github.com/timberio/vector/tree/blt-datadog_logs_perf_investigation | +263 |

The largest contributor is the blackhole sink modification, but do note also that the improvements taken as one are more effective than when taken individually. As of this writing the top contributors to run cycles are:

| Symbol | Cycle % |
| --------- | ----------- |
| `serde_json::read::SliceRead::parse_str_bytes` | 8.55% |
| `core::sync::atomic::atomic_load` | 3.77% | 

At this point I'm satisfied with the improvement on the source side and will adjust to include the sink side. It's unlikely that #8331 will proceed in its current form and the blackhole modification is not appropriate (but happily not relevant shortly) but the parsing improvement and metrics improvements are not contentious.

**Comment by blt** [comment]
Starting from 67bd31c0713da68c11f69eed01dcfa974dfe549d I have modified the vector configuration to the following:

```toml
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_logs"]
default_api_key = "DEADBEEF"
encoding.codec = "json"
endpoint = "http://localhost:8080/v1/input"
```

This now enables the `datadog_logs` sink, sinking into a local [`http_blackhole`](https://github.com/blt/lading/blob/main/lading_blackholes/src/bin/http_blackhole.rs). The `http_gen` configuration is unchanged. The generator and blackhole side are run like so:

```
> firejail --rlimit-as=2g --cpu=0,9 --noprofile ./target/release/http_blackhole --worker-threads 1 --binding-addr "0.0.0.0:8080" --prometheus-addr "0.0.0.0:9002"

> firejail --rlimit-as=8g --cpu=11,12 --noprofile ./target/release/http_gen --config-path configs/http_gen.toml
```

I've introduced the use of [firejail](https://firejail.wordpress.com/) to avoid needing to futz around with cgroups. Previously I ran vector with [taskset](https://linux.die.net/man/1/taskset) but now want to restrict memory consumption as our HTTP sinks have, in the past, continually allocated retries under load. Vector is run like so:

```
> firejail --noprofile --rlimit-as=2g --cpu=1,2,3,4,5,6,7,8 ./target/release/vector --config ~/projects/us/troutwine/lading/configs/vector-ddlogs.toml -qq
```

Note that vector runs on CPUs isolated from its generator/blackhole peers and has a maximum of 2Gb of memory. The results here are poor, which we expect consider that the default values for buffers are very low for this sink. Here are, in order, the views of the `http_gen`, vector and `http_blackhole` bytes throughput for a small window of time:

![Screenshot from 2021-07-19 10-12-55](https://user-images.githubusercontent.com/3579/126199950-f3e57d8d-591a-47e3-9a49-6b22de146154.png)
![Screenshot from 2021-07-19 10-14-25](https://user-images.githubusercontent.com/3579/126200158-c53953c9-f028-4cb8-a546-f05d1fb6e1c4.png)
![Screenshot from 2021-07-19 10-14-53](https://user-images.githubusercontent.com/3579/126200187-8f8d2076-b257-4ad4-9c5b-d83481e70f8b.png)

On the bright side, backpressure to the producer does appear to work. It also appears that the `datadog_logs` sink does not correctly instrument its `component_name` label, note the "Value" field. Throughput into the  blackhole appears to be in the 125Kb/s range as measured from the blackhole, 215Kb/s as measured from vector. The blackhole presently instruments the content length field sent by vector so I will need to confirm whether or not the body is actually the content length size or otheriwse.

**Comment by blt** [comment]
`perf` numbers indicate that `minix_oxide::deflate::core::compress_normal` and `minix_oxide::deflate::DictOxide::find_match` are the most cycle hungry components of vector in this configuration, at 18.1% and 8.64% respectively. 

Setting `compression = "none"` on the sink improves throughput to 195Kb/s -- +70Kb/s -- as measured from the blackhole. Repeating the `perf` run indicates json parsing in the source is not the dominate cycle time, CPU timeline suggests that vector is largely idle:

![Screenshot from 2021-07-19 10-55-26](https://user-images.githubusercontent.com/3579/126205005-4a7217da-ab7d-46a2-b8a6-9a9b2cf6f0fd.png)

Very little CPU time is spent in the sink. I consider it likely but have no data to prove that the upstream pipe into the sink is too small. Note also that `vector_utilization` metric for the egress component is 1.0.

**Comment by blt** [comment]
I have added 

```
buffer.type = "memory" 
buffer.when_full = "drop_newest"
```

to the sink configuration, getting ingress at the generator to ~450Mb/s and egress to the blackhole 195Kb/s. Sink utilization remains 1.0.

**Comment by spencergilbert** [comment]
@blt fyi we closed #7370 which had some background research and thoughts around the sink side of things here, in case it's useful for your work

**Comment by blt** [comment]
I am now re-doing the experiment [here](https://github.com/timberio/vector/issues/8263#issuecomment-882721186) but at 6c24ea82845dfb9f03d6e1e71a80930d0b8cbb3e and with a slight configuration change:

```toml
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_logs"]
default_api_key = "DEADBEEF"
encoding.codec = "json"
endpoint = "http://localhost:8080/v1/input"
batch.max_bytes = 5242880
batch.max_events = 1000 
```

This configuration sets the maximum array size, payload size per https://docs.datadoghq.com/api/latest/logs/#send-logs. Note that vector is configured to only allocate 2Gb. This setup is able to consume 9.3Mb/s of incoming lines prior to malloc failure: vector steadily consumes memory in this setup.

**Comment by blt** [comment]
This experiment duplicates the previous but with ARC flipped on:

```toml
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_logs"]
default_api_key = "DEADBEEF"
encoding.codec = "json"
endpoint = "http://localhost:8080/v1/input"
batch.max_bytes` = 5242880
batch.max_events = 1000 
request.concurrency = "adaptive"
```

In this configuration throughput is 2.3Mb/s with 1 CPU fully pegged. Memory use is stable, if high, at 1Gb.

**Comment by blt** [comment]
Upstream has changed -- notably #8383 -- and will redo the previous experiment at 97c6606383aa3e278216ce309581bd5fe7ef1d97. The configuration is and will be until stated otherwise: 

```
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_logs"]
default_api_key = "DEADBEEF"
encoding.codec = "json"
endpoint = "http://localhost:8080/v1/input"
batch.max_bytes = 5242880
batch.max_events = 1000
```

ARC is left out as it's a complex sub-system and both delayed the allocation pattern seen and regressed throughput. Running in this configuration I find the following:

| SHA | `http_gen` generation Mb/s | `datadog_logs` bytes processed Mb/s | `http_blackhole` receipt Mb/s |
| ------ | ------------------------------------ | -------------------------------------------------- | --------------------------------------- |
| 97c6606383aa3e278216ce309581bd5fe7ef1d97 | 3 | 27 | 20 |
| 6c24ea82845dfb9f03d6e1e71a80930d0b8cbb3e | 2.7 | 28 | 20 | 

The `datadog_logs` throughput has low variability while input and output do. Vector did crash eventually with malloc failure, taking roughly 10 minutes to do so in the case of 97c6606383aa3e278216ce309581bd5fe7ef1d97 and a few minutes for 6c24ea82845dfb9f03d6e1e71a80930d0b8cbb3e. Massif demonstrates that it's tower retries stacking up, for which I have no hypothesis. We do not see a similar issue in our soak infra.

**Comment by blt** [comment]
I've discovered what caused vector to hemorrhage memory. Until https://github.com/blt/lading/commit/1b1a7548c6f716905ce373df85c6ba203721759b the `http_blackhole` would only read the `CONTENT_LENGTH` header from any request. This meant for large bodies the blackhole would hang up before vector finished sending the body, a legit retry opportunity. Unfortunately vector appears to stack up retries until its death. 

With upstream changed -- notably #8383 -- I will redo the previous experiment at 3fd6a53f497f0279024eb5f02c3db2251f7f3f8e. The configuration is and will be until stated otherwise: 

```
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_logs]
type = "datadog_logs"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_logs"]
default_api_key = "DEADBEEF"
encoding.codec = "json"
endpoint = "http://localhost:8080/v1/input"
batch.max_bytes = 5242880
batch.max_events = 1000
```

ARC is left out as it's a complex sub-system and both delayed the allocation pattern seen and regressed throughput. Running in this configuration I find the following:

| SHA | `http_gen` generation Mb/s | `datadog_logs` bytes processed Mb/s | `http_blackhole` receipt Mb/s |
| ------ | ------------------------------------ | -------------------------------------------------- | --------------------------------------- |
|  3fd6a53f497f0279024eb5f02c3db2251f7f3f8e | 10.2 | 10.2 | 7.49 |

Note that generation and processed values are consistent, owing to there being no hang-ups by the blackhole. Vector is configured to compress its output, verified at the blackhole, explaining the smaller output than input. Perf data shows that compression in miniz is 30% of execution instruction, a consequence of `DatadogLogsJsonService::build_request`.

**Comment by blt** [comment]
After some investigation into why perf output showed a preference in vector for work aligned along second boundaries -- see [this comment](https://github.com/timberio/vector/issues/8263#issuecomment-882747157) -- and with a hint from @lukesteensen I've discovered that the default settings for sink rate limiter -- undocumented -- are at fault. Consider the config struct for the sink: https://github.com/timberio/vector/blob/bdb4c6832d9de0bbae59e9c9dafd249c7444322f/src/sinks/datadog/logs/config.rs#L22-L47

Undocumented here is the `request` field. If you trace this back you'll run into https://github.com/timberio/vector/blob/bdb4c6832d9de0bbae59e9c9dafd249c7444322f/src/sinks/util/service.rs#L165-L184

Stepping forward some this feeds https://github.com/timberio/vector/blob/bdb4c6832d9de0bbae59e9c9dafd249c7444322f/src/sinks/util/service.rs#L307-L328

So, the problem we have here is the [layer](https://docs.rs/tower/0.4.8/tower/trait.Layer.html) that controls concurrency -- what we call ARC -- is sat below the rate limiting layer, which by default is 5 requests per second. So, no matter how much concurrency we might detect the endpoint is able to manage we limit, by default, to a fairly conservatively low threshold.

My vector.toml now looks like the following:

```toml
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_agent]
type = "datadog_agent"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_agent"]
endpoint = "http://localhost:8080/v1/input"
default_api_key = "DEADBEEF"

buffer.max_bytes = 10485760
buffer.type = "memory"
buffer.max_events = 2000

batch.max_bytes = 5242880 # https://docs.datadoghq.com/api/latest/logs/#send-logs
batch.max_events = 1000

request.concurrency = 10000
request.rate_limit_num = 10000000
request.rate_limit_duration = 1
```

Note there are a few changes. First we set the buffer max bytes, max events to double the 'batch' size, which controls the payload size sent to Datadog. Send we adjust -- via undocumented knobs -- the request concurrency and rate limit. The main drivers are `rate_limit_num` and `rate_limit_duration`. In the future the default for `concurrency` is intended to be  'adaptive'/ARC but that isn't landed as yet. `http_blackhole` and `http_gen` settings are unchanged. 

With the above vector config we find: 

| SHA | `http_gen` generation Mb/s | `datadog_logs` bytes processed Mb/s | `http_blackhole` receipt Mb/s |
| ------ | ------------------------------------ | -------------------------------------------------- | --------------------------------------- |
|  bdb4c6832d9de0bbae59e9c9dafd249c7444322f | 76.9 | 78.2 | 78.2 |

The configuration adjustment is worth +60Mb/s. Suggestions:

* configure rate limits to be wide open by default, 
* instrument rate limits and their saturation within the tower layer stack discussed and

This work suggests that sink rate limits should be wide-open to start and users that need to throttle down should have the option. This particular bottleneck should be something we can identify easily from telemetry, should such telemetry be present.

**Comment by blt** [comment]
With #8472 and #8471 in master with #8488 soon to be we can simplify the vector config some to 

```toml
data_dir = "/tmp/data/vector"

##
## Sources
##

[sources.internal_metrics]
type = "internal_metrics"

[sources.datadog_agent]
type = "datadog_agent"
acknowledgements = false
address = "0.0.0.0:8282"

##
## Sinks
##

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal_metrics"]
address = "0.0.0.0:9598"

[sinks.dd_logs_egress]
type = "datadog_logs"
inputs = ["datadog_agent"]
endpoint = "http://localhost:8080/v1/input"
default_api_key = "DEADBEEF"

buffer.type = "memory"
buffer.max_bytes = 50000000 # 50 MB
buffer.max_events = 40000   # buffer 40 payloads at a time
```

I settled on a relatively conservative amount of buffer space to feed the now very concurrent sink, 50MB or 40 payloads worth of events, whichever comes first. I don't know that we can set reasonable defaults around buffers as it'll be very contingent on deploy environment. Anyhow. With this change I had to also modify the `http_gen` configuration for more load. It now looks like:

```toml
worker_threads = 2
prometheus_addr = "0.0.0.0:9001"

[targets.simple]
target_uri = "http://localhost:8282/v1/input"
bytes_per_second = "500 Mb"
parallel_connections = 10
method.type = "Post"
method.variant = "DatadogLog"
method.maximum_prebuild_cache_size_bytes = "1 Gb"

[targets.simple.headers]
dd-api-key = "DEADBEEF"
```

Targeting dee27a74f7948f7bd01d09df4ca1944403bc263f from #8488 I find:

| SHA | `http_gen` generation Mb/s | `datadog_logs` bytes processed Mb/s | `http_blackhole` receipt Mb/s |
| ------ | ------------------------------------ | -------------------------------------------------- | --------------------------------------- |
|  dee27a74f7948f7bd01d09df4ca1944403bc263f | 210 | 211 | 211 |

This is +130Mb/s give or take. The primary difference between this run and the last being we avoid path iteration code when we can and avoid serialization between implicit atomics through `lazy_static`. This does suggest that path iteration code in `Event` will be a hotspot where it cannot be avoided, which should be the case through most VRL code. We'll find out soon enough, I imagine. 

Interestingly ~210Mb/s appeared to be a hard limit. Enabling ARC was not able to move past it, giving more buffer space also did nothing. This may be an artifact of my experimental setup or the next scale boundary in vector. Considering that we are now well past our initial goal I consider this worthy of note but not important to track down.

**Comment by jszwedko** [comment]
🚀 🚀 🚀 

Just a small nit that `buffer.max_bytes` doesn't apply for `buffer.type = "memory" so you could simplify the config a bit.
